File: src/main/scala/vm/vagrant/configuration/builder/VagrantProviderConfigBuilder.scala
package vm.vagrant.configuration.builder

import vm.vagrant.configuration.VagrantProviderConfig
import vm.vagrant.configuration.builder.util.VagrantBuilderException

/**
  * Created by oliver.ziegert on 31.03.17.
  */
object VagrantProviderConfigBuilder {
  def create = new VagrantProviderConfigBuilder
}

class VagrantProviderConfigBuilder {
  private var name: String = _
  private var guiMode: Boolean = false
  private var memory: Int = _
  private var cpus: Int = _
  private var customize: List[String] = List()
  private var vmName: String = _

  def withName(name: String): VagrantProviderConfigBuilder = {
    this.name = name
    this
  }

  def withGuiMode(guiMode: Boolean): VagrantProviderConfigBuilder = {
    this.guiMode = guiMode
    this
  }

  def withMemory(memory: Int): VagrantProviderConfigBuilder = {
    this.memory = memory
    this
  }

  def withCpus(cpus: Int): VagrantProviderConfigBuilder = {
    this.cpus = cpus
    this
  }

  def withCustomize(customize: String): VagrantProviderConfigBuilder = {
    this.customize ::= customize
    this
  }

  def withVmName(vmName: String): VagrantProviderConfigBuilder = {
    this.vmName = vmName
    this
  }

  def build(): VagrantProviderConfig = {
    if (name != null && name.isEmpty) throw new VagrantBuilderException("No name defined")
    new VagrantProviderConfig(name = name,
                              guiMode = guiMode,
                              memory = memory,
                              cpus = cpus,
                              customize = customize,
                              vmName = vmName)
  }


}

File: src/main/java/de/oth/clustering/java/vm/vagrant/configuration/builder/VagrantProviderConfigBuilder.java
package de.oth.clustering.java.vm.vagrant.configuration.builder;

import de.oth.clustering.java.vm.vagrant.configuration.VagrantProviderConfig;
import de.oth.clustering.java.vm.vagrant.configuration.builder.util.VagrantBuilderException;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by oliver.ziegert on 20.04.17.
 */
public class VagrantProviderConfigBuilder implements Serializable {
    private String name;
    private Boolean guiMode;
    private Integer memory;
    private Integer cpus;
    private List<String> customize;
    private String vmName;

    public VagrantProviderConfigBuilder() {
        this.customize = new ArrayList<String>();
        this.guiMode = false;
    }

    public static VagrantProviderConfigBuilder create() {
        return new VagrantProviderConfigBuilder();
    }

    public VagrantProviderConfigBuilder withName(String name) {
        this.name = name;
        return this;
    }

    public VagrantProviderConfigBuilder withGuiMode(Boolean guiMode) {
        this.guiMode = guiMode;
        return this;
    }

    public VagrantProviderConfigBuilder withMemory(Integer memory) {
        this.memory = memory;
        return this;
    }

    public VagrantProviderConfigBuilder withCpus(Integer cpus) {
        this.cpus = cpus;
        return this;
    }

    public VagrantProviderConfigBuilder withCustomize(String customize) {
        this.customize.add(customize);
        return this;
    }

    public VagrantProviderConfigBuilder withVmName(String vmName) {
        this.vmName = vmName;
        return this;
    }

    public VagrantProviderConfig build() {
        if (name.isEmpty()) {
            throw new VagrantBuilderException("no name defined");
        }
        return new VagrantProviderConfig(name, guiMode, memory, cpus, customize, vmName);
    }
}
